src/BCFKeyValueData.cc

```diff
@@ -619,7 +618,9 @@ static Status scan_bucket(
                 return Status::IOError("BCFKeyValueData bcf_unpack",
                                        dataset + "@" + query.str());
             }
-            records.push_back(vt);
+            if (predicate == nullptr ||
+                predicate(hdr, vt.get()))
+                records.push_back(vt);
```

Comment:
Let's test the predicate *before* calling `bcf_unpack` -- this will be a significant efficiency gain. However, we should document that the contract for `bcf_predicate` is that the record passed in may not be unpacked. This will give the predicate control over how much to unpack to determine the answer. htslib is nicely designed to permit this progressive unpacking without duplicating work.